compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala

```diff
@@ -458,105 +170,215 @@ class PickleQuotes extends MacroTransform {
-            capturers(ref.symbol)(ref).select(nme.apply).appliedTo(quotes)
-          case tree: RefTree if isCaptured(tree.symbol, level) =>
-            val body = capturers(tree.symbol).apply(tree)
-            if (tree.isType)
-              transformSpliceType(body, body.select(tpnme.Underlying))
-            else
-              val splice = ref(defn.QuotedRuntime_exprSplice).appliedToType(tree.tpe).appliedTo(body)
-              transformSplice(body, splice)
-
-          case tree: DefDef if tree.symbol.is(Macro) && level == 0 =>
-            // Shrink size of the tree. The methods have already been inlined.
-            // TODO move to FirstTransform to trigger even without quotes
-            cpy.DefDef(tree)(rhs = defaultValue(tree.rhs.tpe))
-
-          case tree: DefTree if level >= 1 =>
-            val newAnnotations = tree.symbol.annotations.mapconserve { annot =>
-              val newAnnotTree = transform(annot.tree)(using ctx.withOwner(tree.symbol))
-              if (annot.tree == newAnnotTree) annot
-              else ConcreteAnnotation(newAnnotTree)
-            }
-            tree.symbol.annotations = newAnnotations
-            super.transform(tree)
-          case _ =>
-            super.transform(tree)
-        }
-      }
 
-    private def liftList(list: List[Tree], tpe: Type)(using Context): Tree =
-      list.foldRight[Tree](ref(defn.NilModule)) { (x, acc) =>
-        acc.select("::".toTermName).appliedToType(tpe).appliedTo(x)
-      }
-  }
-}
+  end makeHoles
 
+}
 
 object PickleQuotes {
   import tpd._
 
   val name: String = "pickleQuotes"
 
-  def getLiteral(tree: tpd.Tree): Option[Literal] = tree match {
-    case tree: Literal => Some(tree)
-    case Block(Nil, e) => getLiteral(e)
-    case Inlined(_, Nil, e) => getLiteral(e)
-    case _ => None
-  }
+  def apply(quotes: Tree, body: Tree, contents: List[Tree], originalTp: Type, isType: Boolean)(using Context) = {
+    /** Helper methods to construct trees calling methods in `Quotes.reflect` based on the current `quotes` tree */
+    object reflect extends ReifiedReflect {
+      val quotesTree = quotes
+    }
+
+    /** Encode quote using Reflection.Literal
+      *
+      *  Generate the code
+      *  ```scala
+      *    quotes => quotes.reflect.TreeMethods.asExpr(
+      *      quotes.reflect.Literal.apply(x$1.reflect.Constant.<typeName>.apply(<literalValue>))
+      *    ).asInstanceOf[scala.quoted.Expr[<body.type>]]
+      *  ```
+      *  this closure is always applied directly to the actual context and the BetaReduce phase removes it.
+      */
+    def pickleAsLiteral(lit: Literal) = {
+      val typeName = body.tpe.typeSymbol.name
+      val literalValue =
+        if lit.const.tag == Constants.NullTag || lit.const.tag == Constants.UnitTag then Nil
+        else List(body)
+      reflect.asExpr(body.tpe) {
+        reflect.Literal {
+          reflect.self
+            .select(s"${typeName}Constant".toTermName)
```

Comment:
I would prefer a match on the tag to select the correct symbol

Reply:
Done